### PR TITLE
performance patch4

### PR DESCRIPTION
**Pull request by jd-lara**
This PR substitutes the use of variable names with symbols as keys in the variables containers. It is a large PR and the implementation isn't totally final in order to maintain functionalities. 

Note that at the moment all the key objects for Vars, AuxVars, and IC's equivalent. I only noticed that they COULD be merged into a single object after implementing the VariableKeys and at this state, I am not sure if we will need additional information later so I will keep them separate for the time being. 

There will be another PR with the remaining changes since the size of this PR is growing really large.

### CODE REVIEW

File: src/core/optimization_container.jl [hunk 1]
```diff
@@ -404,55 +380,49 @@ function get_constraint_names(optimization_container::OptimizationContainer)
...
+    axs...;
+    sparse = false,
+)
+    return add_cons_container!(
         optimization_container,
-        make_constraint_name(constraint_type, T),
-        value,
+        cons_type,
+        var_key.entry_type(),
+        var_key.component_type,
```
Comment (daniel-thom) [inline src/core/optimization_container.jl:395]: There is an odd asymmetry the type vs instance. I guess that's just the way it is. Could you pass only `var_key`?

File: src/devices_models/devices/common/add_variable.jl [hunk 2]
```diff
@@ -81,7 +89,7 @@ function add_variable!(
         name = PSY.get_name(d)
         variable[name, t] = JuMP.@variable(
             optimization_container.JuMPmodel,
-            base_name = "$(var_name)_{$(name), $(t)}",
+            # base_name ="$(encode_symbol(D, T))_{$(name), $(t)}",
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/add_variable.jl:92]: I have commented out all the base name kwargs until we decide how to define those for the JuMP model.

File: src/devices_models/devices/common/cost_functions.jl [hunk 3]
```diff
@@ -228,12 +228,12 @@ function pwl_gencost_sos!(
             "Using Piecewise Linear cost function but no variable/parameter ref for ON status is passed. Default status will be set to online (1.0)"
         )
     elseif spec.sos_status == SOSStatusVariable.PARAMETER
-        param_key = encode_symbol(OnVariable, string(spec.component_type))
+        param_key = encode_symbol("ON", string(spec.component_type))
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/cost_functions.jl:231]: No, in a following stage we need to use the parameter keys here

File: src/core/variables.jl [hunk 4]
```diff
@@ -1,50 +1,28 @@
...
-const STOP = "stop"
-const THETA = "theta"
-const VM = "Vm"
-const LIFT = "z"
-const ACTIVE_POWER_PUMP = "Ppump"
-
-abstract type VariableType end
+struct VariableKey{T <: VariableType, U <: PSY.Component} <: OptimizationContainerKey
+    entry_type::Type{T}
+    component_type::Type{U}
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/core/variables.jl:3]: I know is redundant but didn't know what the best way to do it was. I can change the getter function to use this.

File: src/core/optimization_container.jl (hunk 1 above)
Comment (jd-lara) [reply src/core/optimization_container.jl:395]: I asked @kdheepak about this and he pointed out that all the VariableTypes are generally used as instances in the code. I don't have a specific preference.

File: src/core/variables.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/core/variables.jl:3]: based on the @code_lowered results, I think we should keep the duplicity 

image

File: src/core/optimization_container.jl (hunk 1 above)
Comment (jd-lara) [reply src/core/optimization_container.jl:395]: Yes, I will make this change when refactoring the constraints container.

File: src/core/variables.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/core/variables.jl:3]: This post has the answer: https://discourse.julialang.org/t/get-type-of-field-in-parametric-type/8210/2

Define this:
```
findparam(foo::foobar{T, U}) where {T, U} = U
```

[the remark this answers is not in the document]
Comment (jd-lara) [reply src/core/variables.jl:3]: For this case we probably want to define one method for each entry

**Pull request merged**